goa: add optional --debug switch to produce debug build artifacts

mewmew opened the pull request:
Implement debug build mode for the cargo build system.

Follow-up commits are expected to implement debug build mode
for other build systems and programming languages (e.g. CMake
and C/C++).

Also, add rust_debug example program for debugging Rust
programs on Genode.

To run in release build mode, use 'goa run'

To run in debug build mode, use 'goa run --debug'

In the rust_debug demo, wait_for_continue is only invoked in
debug build mode. Furthermore, in debug build mode, full
debug info is included (e.g. Rust debug symbols).

mewmew commented on share/goa/doc/config.txt:
```diff
@@ -139,6 +139,13 @@ Goa evaluates the following configuration variables.
   configuration variable, or the '--license' command-line argument.
 
+:debug:
+  If set to 1, Goa uses debug build mode to produce debug build artifacts,
```
Suggestions for improving the phrasing of this documentation are gladly welcomed.

jschlatow commented:
@mewmew Thank you for opening this PR. I really like the idea of building binaries with debug symbols with Goa.

Looking at the cargo documentation, though, I sense that there are a lot more debug options. There are several variables modified by switching between the _release_ and the _dev_ profile and I have the feeling that mapping all these changes to a single `--debug` flag will not suffice for actual debugging. For instance, I often only need the debug symbols but do not want to change the optimisation level.

My gut feeling is that we could add the same functionality (but with more fine-grained control by the user) if we add support for a _cargo_args_ file (analogously to _make_args_ and _cmake_args_). What do you think?

mewmew commented:
Hi @jschlatow,

Thanks for the feedback!

I definitely agree that adding support for `cargo_args` would be very useful and would mirror the functionality of `make_args` and `cmake_args` in an intuitive way.

That being said, I don't think one has to exclude the other.

Personally, I believe the main strength of Goa is to provide sane default and make *development fun*.

To build a Goa project, just run `goa build`. Regardless of source language and compiler setup.

Want to run a Goa project, just run `goa run`.

Want to publish a Goa project, just run `goa publish`.

Having something along the lines of `goa debug` would make sense, as it is such a common development use-case.

And not having to read up on `what make_args`, `cmake_args` and `cargo_args` to use to get a sane default for debugging would be a huge benefit to the "joy of coding".

Since wanting to include debug symbols is such a common wish when debugging, having a `--debug` switch in Goa that is implemented by *all supported programming language and build systems* makes the Goa build system really joyful to use.

Another aspect is that it is very common to want to switch between release builds and debug builds. And having a simple flag to toggle that is ideal. If we had say a `cargo_args` flag in the root of the repository, then we'd have to either temporarily remove the file or edit it to comment out various arguments to switch between release and debug builds.

So, in favour of "joyful development" I would very much favour having both support for a `--debug` switch in Goa, and having support for more fine-grained `cargo_args` when deemed necessary (to e.g. add debug symbols but still use `-O2`).

Would be happy to continue the discussion : )

Cheers,
Robin

atopia reviewed (changes requested):
Thanks a lot for your contribution, apart from a few nitpicks this looks good to me.

jschlatow commented:
@atopia Thanks for your review.

@mewmew I very much agree with your line of thought. I would love to see a `goa debug` command in the future. Would you mind documenting the cargo-specific implications of the `--debug` flag in the build-systems.txt. I think it could be helpful for developers to know that it switches to the _dev_ profile.

mewmew commented on examples/rust_debug/src/build.rs:
```diff
@@ -0,0 +1,6 @@
+fn main() {
+    // link to ld.lib.so in debug build mode (needed for wait_for_continue).
+    #[cfg(debug_assertions)] {
+        println!("cargo:rustc-link-arg=-l:ld.lib.so");
```
> While the idea is good in principle, the build script is not necessary, because ld.lib.so is already included as a symbol of the base dependency.

Oh, that's good to know! Will help me simplify a few things on my end as well : ) Thanks!

jschlatow commented:
@mewmew While testing your changes, I noticed that the binary is not automatically rebuilt when changing between `goa build` and `goa build --debug`. Do you have an idea how this could be achieved?

jschlatow commented:
@mewmew I merged your commits to staging with adapted commit titles (to fit the recommended limit of 50 characters). I also added a few minor style fixups and added support for the `--debug` switch to make, cmake and qmake.

mewmew commented:
> @mewmew I merged your commits to staging with adapted commit titles (to fit the recommended limit of 50 characters). I also added a few minor style fixups and added support for the --debug switch to make, cmake and qmake.

Wonderful! Thanks a lot @jschlatow and @atopia for reviewing the PR and improving it : ) I also learnt a lot (e.g. no need to link against `ld.lib.so` when including `base` in `used_api`).

Cheerful regards and happy coding (and debugging : ),
Robin
